draft-ietf-quic-transport.md

```diff
@@ -2178,12 +2180,16 @@ before the handshake is confirmed, as defined in section 4.1.2 of {{QUIC-TLS}}.
 
 If the peer sent the disable_active_migration transport parameter, an endpoint
 also MUST NOT send packets (including probing packets; see {{probing}}) from a
-different local address to the address the peer used during the handshake. An
-endpoint that has sent this transport parameter, but detects that a peer has
-nonetheless migrated to a different remote address MUST either drop the incoming
-packets on that path without generating a stateless reset or proceed with path
-validation and allow the peer to migrate. Generating a stateless reset or
-closing the connection would allow third parties in the network to cause
+different local address to the address the peer used during the handshake. If
+the server sets the Disable Migration field in the preferred_address transport
+parameter, the client MUST NOT send packets from a different local address to
+the server's preferred address.
```

Comment:
I think it might make sense to drop this MUST NOT as a whole:
* What if SPA includes an address of a different address family than the original address? The client has to use a different local address.
* A QUIC stack might be letting the operating system select the outgoing address, by skipping the call to `bind`, and letting `connect` select the *source* address. I do not think we'd want to forbid such QUIC stacks not supporting SPA.

FWIW, I do not think there is a big problem in client switching to a different address when migrating the preferred address with migration disabled. If the path is established, then there's no problem. If the path fails to establish, endpoints can continue talking on the original path.

Reply:
@kazuho, it is a good point about a different address family.

> * What if SPA includes an address of a different address family than the original address? The client has to use a different local address.

If a server does not want client to change addresses, which it must do, if changing IP families, it will just not supply the alternative IP family SPA, so the Disable Migration flag can apply only to the same family as the handshake address.  Or each family may get its own flag, which increases flexibility.

> * A QUIC stack might be letting the operating system select the source address, by skipping the call to bind, and letting connect select that address. I do not think we'd want to forbid such QUIC stacks not supporting SPA.

First, it is only SPAs with Disable Migration flag that are at issue here.  In any case, I am ok with all s/MUST/SHOULD/g in regard to disabling migration, since unintentional migration happens, which can lead to poor performance for the client.  In case of the client that does not bind, Disable Migration is a hint, and the client can always try harder to follow it, since it is in its interests.  That client can, for example, actually bind another socket to the address selected by the kernel and use that socket for the subsequent communication w/ the server.

> FWIW, I do not think there is a big problem in client switching to a different address when migrating the preferred address with migration disabled. If the path is established, then there's no problem. If the path fails to establish, endpoints can continue talking on the original path.

It is a problem (big or small is for the client to decide), if the server is chosen based on the client's original network attachment, and the client decides to migrate to a network provider that is inappropriate for the chosen server (during SPA migrations or not).  If the client tries, and the path fails to establish, the client wasted an rtt or possibly some timeout.  Or the path may actually get established, but the performance will be poor.  The client is better off quickly re-establishing a new connection.

Reply:
> First, it is only SPAs with Disable Migration flag that are at issue here.

Based on the discussion on the issue (e.g., https://github.com/quicwg/base-drafts/issues/3765#issuecomment-657903203), I think we tend to agree that the problem exists to the same extent, for servers disabling migration on SPA and those do not. This is essentially a server asking the client to migrate to a specific server address based on the address that the client is using, and therefore not wanting the client to move to a different client address.

Therefore, if we are going to have any requirement, I think we should recommend *one* behavior, regardless of the use of the disable-migration flag.

Having different behaviors might seem like a way of restoring the old behavior (I have reservations as stated in https://github.com/quicwg/base-drafts/issues/3765#issuecomment-657925719). But regardless, I would argue that having different behavior is an unnecessary complexity without technical merit.